{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# RNA-Seq Pipeline\n",
    "## Step 3: Differential Expression Analysis\n",
    "\n",
    "This notebook performs differential expression analysis using:\n",
    "- Normalized count matrix\n",
    "- A sample metadata table\n",
    "- log2 fold change\n",
    "- Adjusted p-values\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from scipy.stats import ttest_ind"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load normalized counts (TPM or filtered counts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tpm_matrix = pd.read_csv('../results/tpm_normalized_matrix.csv', index_col=0)\n",
    "tpm_matrix.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load metadata (manually created or downloaded)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata = pd.read_csv('../data/metadata.csv')\n",
    "metadata"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Split samples by condition"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "group1 = metadata[metadata['condition'] == 'control']['sample']\n",
    "group2 = metadata[metadata['condition'] == 'treatment']['sample']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Differential Expression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "results = []\n",
    "for gene in tpm_matrix.index:\n",
    "    x = tpm_matrix.loc[gene, group1]\n",
    "    y = tpm_matrix.loc[gene, group2]\n",
    "    \n",
    "    lfc = np.log2((y.mean() + 1) / (x.mean() + 1))\n",
    "    pval = ttest_ind(x, y, equal_var=False).pvalue\n",
    "    \n",
    "    results.append([gene, lfc, pval])\n",
    "\n",
    "deg_df = pd.DataFrame(results, columns=['gene', 'log2FC', 'pvalue'])\n",
    "\n",
    "# FDR correction\n",
    "deg_df['padj'] = deg_df['pvalue'] * len(deg_df) / (deg_df['pvalue'].rank())\n",
    "\n",
    "deg_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Save results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "deg_df.to_csv('../results/differential_expression_results.csv', index=False)\n",
    "deg_df"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
